### PR TITLE
fix(globset/serde1): allow deserializing owned strings into `Glob`

### LINKED ISSUE #2386: globset: `serde` feature - cannot deserialize `String` into `Glob`
Hi, I hope that posting issues about subcrates is ok.

#### What version of globset are you using?

`globset = { version = "0.4.9", features = ["serde1"] }`.

#### Describe your bug.

Because the implementation of trait `Deserialize` uses `&str`, it does not accept deserializing owned `String`.

https://github.com/BurntSushi/ripgrep/blob/61101289fabc032fd8e90009c41d0b78e6dfc9a2/crates/globset/src/serde_impl.rs#L15-L22

#### What are the steps to reproduce the behavior?

This problem occurred to me when I tried to parse `Glob`s from a `toml::Value` object (already obtained from a file). As `toml::Value` owns strings, it caused a problem.

If needed, I can include a MWE (but quite large). 

#### What is the actual behavior?

When parsing some globs, here is the error obtained:

```
Error: TomlDecode(Error { inner: ErrorInner { kind: Custom, line: None, col: 0, at: None, message: "invalid type: string \"src/**.rs\", expected a borrowed string", key: [] } })
```

#### What is the expected behavior?

Not sure if it is interesting performance-wise, but using a custom `deserialize_with` function, with `String` instead of `&str`, fixes the problem (`glob.as_str()`) has to be used afterward).

I can provide tests and make a PR if you are ok with this solution. Otherwise, I am also open to discussions :-)

### PR DESCRIPTION
**Pull request by jeertmans**
This PR modifies how `Glob` is deserialized such that it now allows for owned strings (previously, only borrowed) to be deserialized into `Glob`. This takes inspiration from what is done in the [`regex_serde` crate](https://github.com/tailhook/serde-regex/blob/336bb456ecd146ba9e3fcc2fef71870f603d72c5/src/lib.rs#L179-L191).

For more details, see the discussion in #2386.

The first commit includes tests, which fail without the solution.
The second commit include the solution.

This fixes #2386.

I'd be happy to hear from your thoughts and I'm open to any review :-)

### CODE REVIEW

**Comment by jeertmans** [comment]
Like so @LPGhatguy? I have updated the code according to your comment.

I have merged `StrVisitor` and `StringVisitor` implementation into `CowStrVisitor` (see [here](https://rust-random.github.io/rand/src/serde/de/impls.rs.html#599)).

After checking, the first test uses a `Cow::Owned`, while the two others use a `Cow::Borrowed`.

**Comment by LPGhatguy** [comment]
`Cow` doesn't need to be involved here. You should make a `GlobVisitor` that constructs the glob directly from the value instead of pulling it into an intermediate `Cow`.

I don't think that `visit_borrowed_bytes` or `visit_byte_buf` will be reached. Because we're calling `deserialize_str`, we're telling Sede that we're going to deserialize a string, not an arbitrary sequence of bytes.

**Comment by jeertmans** [comment]
Thanks, @LPGhatguy! Indeed, after my commit, I realized that using `Cow` was no longer needed. I implemented a new version where I directly parse a string into a `Glob`.

File: crates/globset/src/serde_impl.rs [hunk 1]
```diff
@@ -12,18 +13,62 @@ impl Serialize for Glob {
     }
 }
 
+struct GlobVisitor;
+
+impl<'a> Visitor<'a> for GlobVisitor {
+    type Value = Glob;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a string or a borrowed string")
```
Comment (LPGhatguy) [inline crates/globset/src/serde_impl.rs:22]: This should probably be "a glob pattern" or something more descriptive here.

**Review by LPGhatguy — COMMENTED** [review]
Last comment, then LGTM.

File: crates/globset/src/serde_impl.rs (hunk 1 above)
Comment (jeertmans) [reply crates/globset/src/serde_impl.rs]: Thanks for noticing! There you go :)

**Review by jonasbb — APPROVED** [review]

**Review by LPGhatguy — APPROVED** [review]
lgtm

**Comment by jeertmans** [comment]
Hello @BurntSushi, any plan on merging this? Or is it waiting for something else?

File: crates/globset/src/serde_impl.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
-use serde::de::Error;
+use serde::de::{Error, Visitor};
 use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt;
```
Comment (BurntSushi) [inline crates/globset/src/serde_impl.rs:3]: Please group imports. std imports should come first, then an empty line, then third party imports (like `serde`) and then finally crate local imports.

Comment (BurntSushi) [reply crates/globset/src/serde_impl.rs:3]: Or just write `std::fmt::Formatter` and `std::fmt::Result` below.

File: crates/globset/src/serde_impl.rs [hunk 3]
```diff
@@ -12,18 +13,62 @@ impl Serialize for Glob {
...
-        Glob::new(glob).map_err(D::Error::custom)
+        deserializer.deserialize_str(GlobVisitor)
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use Glob;
+    use crate::glob::Glob;
+    use std::collections::HashMap;
```
Comment (BurntSushi) [inline crates/globset/src/serde_impl.rs:44]: Same nit here about import grouping.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
I think this LGTM and looks much better than what it was before. There are some style nits though.

**Comment by BurntSushi** [comment]
> Or is it waiting for something else?

It's waiting on me to review it. In general, my review latency is pretty high.

File: crates/globset/src/serde_impl.rs (hunk 2 above)
Comment (jeertmans) [reply crates/globset/src/serde_impl.rs]: Like so (see recent commit)?

**Comment by jeertmans** [comment]
> It's waiting on me to review it. In general, my review latency is pretty high.

No problem, take your time :-)

I think I have updated the code accordingly.

**Review by BurntSushi — APPROVED** [review]
LGTM

**Comment by BurntSushi** [comment]
In the future, for changes this small, they should be one commit. If you need to make changes after opening the PR, then modify the commit via `git commit --amend` and force push.

Basically, treat commits like code. Optimize for the reader.

**Pull request closed without merging**